installers/macos/preinstall

```diff
@@ -27,5 +34,76 @@ then
 	$INSTALL_LOCATION/scripts/service_stop
 fi
 
-exit 0
+# We need to ensure that there is a psuedo user (aka role) account for
+# our daemon, so that the daemon runs with minimal permissions and not
+# as the usual logged in user.  This is similar to the "nobody"
+# account.
+
+dscl . read /Users/$USERNAME >/dev/null 2>/dev/null
+if [ $? -eq 0 ]
+then
+	echo "Using existing pseudo user $USERNAME"
+	exit 0
+fi
+
+# We need to create a new pseudo user using a "role account".  These
+# have UIDs in the 200-400 range.
+#
+# Apple does not give us any tools to find or request an unused UID,
+# so we have to search for an unused UID.
+#
+# Find all of the already used role account UIDs and exclude them from
+# the allowable range. Build the free set backwards, so that we're not
+# as likely to collide with well-known services (that may not
+# currently be installed).
+
+USED_IDS=$(dscl . list /Users UniqueID | awk '{ if ($2 >= 200 && $2 < 400) print $2}' | sort -n)
+FREE_IDS=""
+for k in {200..399}
+do
+	echo $USED_IDS | grep -v -q $k && FREE_IDS="$k $FREE_IDS"
+done
+
+# We probably only need to try to create the pseudo user using the
+# first free UID in our list, but lets try others if we get an error.
+#
+# BTW `sysadminctl -addUser ...` does not set an exit code if our UID
+# guess is already in use, so we need to check the user database to
+# see if it was actually created.
+#
+# However, 'sysadminctl' does set an exit code if we are not running
+# as root.
```

Comment:
I assume this means it always returns `"0"` if running as root?

Reply:
yeah, their tools are somewhat lacking. we get a 56 exit code if we don't have root access, but other wise it seems it always returns 0 whether it does anything or not... (i'll revisit the wording in the morning after it has had a chance to settle.)